src/Extensions/ElementalAreasExtension.php

```diff
@@ -229,13 +229,26 @@ public function onBeforeWrite()
 
         $ownerClassName = get_class($this->owner);
 
-        // Update the OwnerClassName on EA if the class has changed
         foreach ($elementalAreaRelations as $eaRelation) {
+            // Update the OwnerClassName on EA if the class has changed
+            /** @var ElementalArea $ea */
             $ea = $this->owner->$eaRelation();
             if ($ea->OwnerClassName !== $ownerClassName) {
                 $ea->OwnerClassName = $ownerClassName;
                 $ea->write();
             }
+
+            // Owner page will have set its own HasBrokenLink / HasBrokenFile in SiteTreeLinkTracking / FileLinkTracking
+            // on any $page->Content BEFORE we get to this point
+            /** @var BaseElement $element */
+            foreach ($ea->Elements() as $element) {
+                if ($element->hasField('HasBrokenLink') && $element->HasBrokenLink) {
+                    $this->owner->HasBrokenLink = true;
+                }
+                if ($element->hasField('HasBrokenFile') && $element->HasBrokenFile) {
+                    $this->owner->HasBrokenFile = true;
+                }
+            }
```

Comment:
Did some rough profiling on this with Xdebug enabled, with 100 Elements on a Page it didn't seem make too big a difference in response time (less than .5s), but could scale poorly in more complex situations (e.g. when Layout Blocks are used.) To minimise the overhead you could remove the loop and use `in_array()` with `column()` instead:

```php
            if (in_array(1, $ea->Elements()->column('HasBrokenLink'))) {
                $this->owner->HasBrokenLink = true;
            }

            if (in_array(1, $ea->Elements()->column('HasBrokenFile'))) {
                $this->owner->HasBrokenFile = true;
            }
```

This seemed to work out to about a 75% reduction in processing time in my profiling. (`Elements()` caches after the first fetch.)

Reply:
Actually, I just realised there's no provision here for removing the HasBrokenLink state from the Page, so the following might make more sense (though it implies that Elements are the single source of truth for these properties):

```php
            $this->owner->HasBrokenLink = in_array(1, $ea->Elements()->column('HasBrokenLink'));
            $this->owner->HasBrokenFile = in_array(1, $ea->Elements()->column('HasBrokenFile'));
```

Reply:
Has to be the first suggestion, otherwise if HasBrokenLink/File was `true` before this extension, such as a broken like on $page->Content, the extension will incorrectly override that.

We don't need to worry about setting it to `false` because SiteTree will default it to `false` [here](https://github.com/silverstripe/silverstripe-cms/blob/4/code/Model/SiteTreeLinkTracking.php#L118) before looking for anything that's broken then setting it to `true`